### ISSUE TITLE
Sluggishness when adding polylines

### ISSUE THREAD

**Issue by TomPed** [AnalyticalGraphicsInc/cesium#3740]
I've compared three ways of adding polylines to Cesium to try and narrow down why adding polylines slows down Cesium in a few cases.

## One

Adding polylines using `viewer.entities.add`. From the `.gif` below you can see that there is a slow down (Cesium freezes up) when adding the polylines. After the first few polylines added the slow down becomes very, very small, but it's still noticeable.

![addpolylineone](https://cloud.githubusercontent.com/assets/11602092/13932075/d85d34c4-ef7c-11e5-8fe5-1055f7e91013.gif)

```javascript
viewer.entities.add({
    polyline : {
        positions : [cartesian, Cesium.Cartesian3.fromDegrees(longitude, latitude, altitude + 9.5)],
        width : 2
    }
});
```

[Full code](https://gist.github.com/TomPed/5e61a4a997919545e299)

## Two

Adding polylines using `viewer.entities.add`, however, I am using a `CallbackProperty` for the polyline's position. From the `.gif` below you can see there is no slow down at all.

![addpolylinetwo](https://cloud.githubusercontent.com/assets/11602092/13932080/dbb542c4-ef7c-11e5-8c76-ee90a915cfe0.gif)

```javascript
viewer.entities.add({
    polyline : {
        positions : new Cesium.CallbackProperty(function() {
            return [cartesian, Cesium.Cartesian3.fromDegrees(longitude, latitude, altitude + 9.5)];
        }, false),
        width : 2
    }
});
```

[Full code](https://gist.github.com/TomPed/3e667a02ec1508dc2019)

## Three

Adding polylines using Cesium geometries and `Primitive`. From the `.gif` below you might not be able to tell but the first polyline added takes a few milliseconds to get added (Cesium does not freeze up like in the first way of adding polylines). After the first polyline there are no more slowdowns.

![addpolylinethree](https://cloud.githubusercontent.com/assets/11602092/13932084/e04abfe4-ef7c-11e5-8b83-57d80e80b024.gif)

```javascript
scene.primitives.add(new Cesium.Primitive({
    geometryInstances : new Cesium.GeometryInstance({
        geometry : new Cesium.PolylineGeometry({
            positions : [cartesian, Cesium.Cartesian3.fromDegrees(longitude, latitude, altitude + 9.5)],
            width : 2.0,
            vertexFormat : Cesium.PolylineColorAppearance.VERTEX_FORMAT,
            followSurface: false
        }),
        attributes: {
            color: Cesium.ColorGeometryInstanceAttribute.fromColor(Cesium.Color.WHITE)
        }
    }),
    appearance : new Cesium.PolylineColorAppearance()
}));
```

[Full code](https://gist.github.com/TomPed/95215cf0fdcf27a4eb70)

**Comment by pjcozzi** [comment]
Thanks for the detailed report.

For **Two**, it's possible that the shader cache is getting trashed (however, Cesium waits like 120 frames before deleting a shader to avoid a case like this) or re-batching is doing something really slow.

No rush, but @mramato can investigate.  If it is something low level, I will take a look.

**Comment by bagnell** [comment]
For three, the first line might be slow because it is starting up the web workers. Try adding `asynchronous : false` as an option parameter to `Primitive`.

**Comment by TomPed** [comment]
You're right, adding `asynchronous : false` to `Primitive` in __Three__ does make the slowdown go away.

**Comment by bagnell** [comment]
If you bring the `if` block [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Primitive.js#L771) out to where `createGeometryTaskProcessors ` is declared, it should fix __Three__ and might fix the others.

**Comment by mramato** [comment]
Good call @bagnell, the worker startup time has annoyed me forever and moving it to initial loading of Primitive.js should do the trick
